src/Templates/BlockTemplatesCompatibility.php

```diff
@@ -0,0 +1,277 @@
+	 *
+	 * @return array
+	 */
+	public function update_render_block_data( $parsed_block, $source_block, $parent_block ) {
+
+		if ( ! $this->is_archive_template() ) {
+			return $parsed_block;
+		}
+
+		/**
+		 * Custom data can be injected to top level block only, as Gutenberg
+		 * will use this data to render the blocks and its nested blocks.
+		 */
+		if ( $parent_block ) {
+			return $parsed_block;
+		}
+
+		array_walk( $parsed_block['innerBlocks'], array( $this, 'inner_blocks_walker' ) );
+
+		return $parsed_block;
+	}
+
+	/**
+	 * Inject hooks to rendered content of corresponding blocks.
+	 *
+	 * @param mixed $block_content The rendered block content.
+	 * @param mixed $block         The parsed block data.
+	 * @return string
+	 */
+	public function inject_hooks( $block_content, $block ) {
+		if ( ! $this->is_archive_template() ) {
+			return $block_content;
+		}
+		/**
+		 * If the block is not inherited, we don't need to inject hooks.
+		 */
+		if ( empty( $block['attrs']['isInherited'] ) ) {
+			return $block_content;
+		}
+
+		$block_name = $block['blockName'];
+
+		/**
+		 * The core/post-template has two different block names:
+		 * - core/post-template when the wrapper is rendered.
+		 * - core/null when the loop item is rendered.
+		 */
+		if (
+			'core/null' === $block_name &&
+			isset( $block['attrs']['__woocommerceNamespace'] ) &&
+			'woocommerce/product-query/product-template' === $block['attrs']['__woocommerceNamespace']
+		) {
+			$block_name = self::LOOP_ITEM_ID;
+		}
+
+		if ( ! in_array( $block_name, array_keys( $this->hook_data ), true ) ) {
+			return $block_content;
+		}
+
+		$hooks = $this->hook_data[ $block_name ];
+
+		return sprintf(
+			'%1$s%2$s%3$s',
+			$this->get_hooks_buffer( $hooks, 'before' ),
+			$block_content,
+			$this->get_hooks_buffer( $hooks, 'after' )
+		);
+	}
+
+	/**
+	 * Get the buffer content of the hooks to append/prepend to render content.
+	 *
+	 * @param array  $hooks    The hooks to be rendered.
+	 * @param string $position The position of the hooks.
+	 *
+	 * @return string
+	 */
+	protected function get_hooks_buffer( $hooks, $position ) {
+		ob_start();
+		foreach ( array_keys( $hooks ) as $hook ) {
+			if ( $hooks[ $hook ]['position'] === $position ) {
+				do_action( $hook );
+			}
+		}
+		return ob_get_clean();
+	}
+
+	/**
+	 * Loop through inner blocks recursively to find the Products blocks that
+	 * inherits query from template.
+	 *
+	 * @param array $block Parsed block data.
+	 */
+	protected function inner_blocks_walker( &$block ) {
+		if (
+			'core/query' === $block['blockName'] &&
+			isset( $block['attrs']['namespace'] ) &&
+			'woocommerce/product-query' === $block['attrs']['namespace'] &&
+			isset( $block['attrs']['query']['inherit'] ) &&
+			$block['attrs']['query']['inherit']
+		) {
+			$block['attrs']['isInherited'] = 1;
+			array_walk( $block['innerBlocks'], array( $this, 'inject_attribute' ) );
```

Comment:
Newbie PHP question: Considering we are already setting `isInherited` inside `inject_attribute`, would this be equivalent?
```suggestion
			$this->inject_attribute( $block );
```

Reply:
Excellent! Thanks for the catch.